File: binom_assistant/interfaces/web/static/js/modules/performance_segmenter.js
/**
 * Модуль: Сегменты эффективности
 * Разделение кампаний на топ/средние/слабые
 */
(function() {
    const PerformanceSegmenterModule = {
        id: 'performance_segmenter',

        translations: {
            total_campaigns: 'Всего кампаний',
            stars_count: 'Звезды (топ 25%)',
            performers_count: 'Перформеры (25-50%)',
            average_count: 'Середнячки (50-75%)',
            underperformers_count: 'Аутсайдеры (нижние 25%)'
        },

        algorithm: `
            <ol>
                <li>Загрузка данных за последние N дней</li>
                <li>Фильтрация кампаний:
                    <ul>
                        <li>Средний расход >= min_daily_spend (по умолчанию $1/день)</li>
                        <li>Минимум min_clicks кликов за период (по умолчанию 50)</li>
                    </ul>
                </li>
                <li>Расчет метрик для каждой кампании:
                    <ul>
                        <li>ROI = (revenue - cost) / cost * 100</li>
                        <li>CR = leads / clicks * 100</li>
                        <li>Profit = revenue - cost</li>
                    </ul>
                </li>
                <li>Вычисление квартилей (перцентилей):
                    <ul>
                        <li>Q1 (25й перцентиль)</li>
                        <li>Q2 (50й перцентиль, медиана)</li>
                        <li>Q3 (75й перцентиль)</li>
                    </ul>
                </li>
                <li>Сегментация кампаний по ROI:
                    <ul>
                        <li><strong>Звезды:</strong> ROI > Q3 (топ 25%)</li>
                        <li><strong>Перформеры:</strong> Q2 < ROI <= Q3 (25-50%)</li>
                        <li><strong>Середнячки:</strong> Q1 < ROI <= Q2 (50-75%)</li>
                        <li><strong>Аутсайдеры:</strong> ROI <= Q1 (нижние 25%)</li>
                    </ul>
                </li>
                <li>Сортировка кампаний внутри каждого сегмента по ROI (убывание)</li>
                <li>Расчет агрегированных метрик для каждого сегмента</li>
            </ol>
        `,

        metrics: `
            <li><strong>Total Campaigns</strong> - общее количество проанализированных кампаний</li>
            <li><strong>Q1, Q2, Q3</strong> - значения квартилей ROI</li>
            <li><strong>Segment Count</strong> - количество кампаний в сегменте</li>
            <li><strong>Total Cost</strong> - общий расход сегмента</li>
            <li><strong>Total Revenue</strong> - общий доход сегмента</li>
            <li><strong>Total Profit</strong> - общая прибыль сегмента</li>
            <li><strong>Avg ROI</strong> - средний ROI сегмента</li>
        `,

        paramTranslations: {
            days: 'Период анализа (дней)',
            min_daily_spend: 'Минимальный расход в день ($)',
            min_clicks: 'Минимум кликов'
        },

        renderTable: function(results, container) {
            if (!results.data || !results.data.segments) {
                container.innerHTML = '<p class="text-muted">Нет данных для отображения</p>';
                return;
            }

            const segments = results.data.segments;
            const period = results.data.period || {};
            const params = results.data.params || {};
            const quartiles = results.data.quartiles || {};

            if (segments.length === 0) {
                container.innerHTML = '<p class="text-muted">Сегментов не найдено. Попробуйте изменить параметры.</p>';
                return;
            }

            let html = '';

            segments.forEach(segment => {
                if (segment.campaign_count === 0) {
                    return; // Пропускаем пустые сегменты
                }

                const profitClass = segment.total_profit >= 0 ? 'text-success' : 'text-danger';
                let segmentColor = '#1a1a1a';
                let borderColor = '#444';

                // Цвета для разных сегментов
                switch(segment.segment_id) {
                    case 'stars':
                        borderColor = 'rgba(75, 192, 192, 0.8)';
                        break;
                    case 'performers':
                        borderColor = 'rgba(54, 162, 235, 0.8)';
                        break;
                    case 'average':
                        borderColor = 'rgba(255, 206, 86, 0.8)';
                        break;
                    case 'underperformers':
                        borderColor = 'rgba(255, 99, 132, 0.8)';
                        break;
                }

                html += `
                    <div class="segment-block" style="margin: 20px 0; padding: 15px; border: 2px solid ${borderColor}; border-radius: 8px; background: ${segmentColor};">
                        <h4>${segment.segment_name}</h4>
                        <p class="text-muted">${segment.description}</p>
                        <div class="segment-summary" style="display: grid; grid-template-columns: repeat(auto-fit, minmax(150px, 1fr)); gap: 10px; margin: 10px 0;">
                            <div>
                                <strong>Кампаний:</strong> ${segment.campaign_count}
                            </div>
                            <div>
                                <strong>Средний ROI:</strong> <span class="${segment.avg_roi >= 0 ? 'text-success' : 'text-danger'}">${segment.avg_roi.toFixed(1)}%</span>
                            </div>
                            <div>
                                <strong>Общий расход:</strong> ${formatCurrency(segment.total_cost)}
                            </div>
                            <div>
                                <strong>Общий доход:</strong> ${formatCurrency(segment.total_revenue)}
                            </div>
                            <div>
                                <strong>Прибыль:</strong> <span class="${profitClass}">${formatCurrency(segment.total_profit)}</span>
                            </div>
                        </div>
                        <div class="table-container" style="margin-top: 15px;">
                            <table>
                                <thead>
                                    <tr>
                                        <th>Кампания</th>
                                        <th>Клики</th>
                                        <th>Расход</th>
                                        <th>Доход</th>
                                        <th>Прибыль</th>
                                        <th>ROI (%)</th>
                                        <th>CR (%)</th>
                                        <th>Средн. расход/день</th>
                                        <th>Binom</th>
                                    </tr>
                                </thead>
                                <tbody>
                `;

                segment.campaigns.forEach(campaign => {
                    const binomId = campaign.binom_id || campaign.campaign_id;
                    const roiClass = campaign.roi >= 0 ? 'text-success' : 'text-danger';
                    const profitClass = campaign.profit >= 0 ? 'text-success' : 'text-danger';

                    html += `
                        <tr>
                            <td>
                                <strong>[${binomId}] ${escapeHtml(campaign.name)}</strong><br>
                                <small class="text-muted">${escapeHtml(campaign.group)}</small>
                            </td>
                            <td>${formatNumber(campaign.total_clicks)}</td>
                            <td>${formatCurrency(campaign.total_cost)}</td>
                            <td>${formatCurrency(campaign.total_revenue)}</td>
                            <td><span class="${profitClass}">${formatCurrency(campaign.profit)}</span></td>
                            <td><span class="${roiClass}">${campaign.roi.toFixed(1)}%</span></td>
                            <td>${campaign.cr.toFixed(2)}%</td>
                            <td>${formatCurrency(campaign.avg_daily_spend)}</td>
                            <td>${renderBinomLink(binomId)}</td>
                        </tr>
                    `;
                });

                html += `
                                </tbody>
                            </table>
                        </div>
                    </div>
                `;
            });

            // Info banner в конце
            html += `
                <div class="info-banner">
                    <strong>Период:</strong> ${period.days || 7} дней |
                    <strong>Мин. расход:</strong> ${params.min_daily_spend || 1}$/день |
                    <strong>Мин. кликов:</strong> ${params.min_clicks || 50} |
                    <strong>Квартили ROI:</strong> Q1=${quartiles.q1}% | Q2=${quartiles.q2}% | Q3=${quartiles.q3}%
                </div>
            `;

            container.innerHTML = html;
        }
    };

    // Регистрируем модуль
    if (typeof ModuleRegistry !== 'undefined') {
        ModuleRegistry.register(PerformanceSegmenterModule);
    }
})();

File: binom_assistant/modules/segmentation/performance_segmenter.py
"""
Модуль сегментации по эффективности (Performance Segmenter)
"""
from typing import Dict, Any, List
from datetime import datetime, timedelta
from sqlalchemy import func
from contextlib import contextmanager
from collections import defaultdict
import statistics

from storage.database.base import get_session
from storage.database.models import Campaign, CampaignStatsDaily
from ..base_module import BaseModule, ModuleMetadata, ModuleConfig


@contextmanager
def get_db_session():
    """
    Локальная обертка над get_session() для использования в with.
    Преобразует генератор в контекстный менеджер.
    """
    session_gen = get_session()
    session = next(session_gen)
    try:
        yield session
    finally:
        try:
            next(session_gen)
        except StopIteration:
            pass


class PerformanceSegmenter(BaseModule):
    """
    Сегментация кампаний по эффективности.

    Разделение кампаний на топ/средние/слабые для применения разных стратегий управления.
    Автоматическое распределение по квартилям.

    Логика сегментации:
    - Топ 25%: звезды (ROI > 75й перцентиль)
    - 25-50%: перформеры (ROI между медианой и 75й перцентиль)
    - 50-75%: середнячки (ROI между 25й перцентиль и медианой)
    - Нижние 25%: аутсайдеры (ROI < 25й перцентиль)
    - Минимальный расход > min_daily_spend (по умолчанию $1/день)
    """

    def get_metadata(self) -> ModuleMetadata:
        """Возвращает метаданные модуля"""
        return ModuleMetadata(
            id="performance_segmenter",
            name="Сегменты эффективности",
            category="segmentation",
            description="Разделение кампаний на топ/средние/слабые",
            detailed_description="Сегментирует все кампании по эффективности для применения разных стратегий управления. Автоматическое распределение по квартилям ROI.",
            version="1.0.0",
            author="Binom Assistant",
            priority="medium",
            tags=["segmentation", "performance", "roi", "quartiles"]
        )

    def get_default_config(self) -> ModuleConfig:
        """Возвращает конфигурацию по умолчанию"""
        return ModuleConfig(
            enabled=True,
            schedule="",  # Некритический модуль - не запускать автоматически
            alerts_enabled=False,  # Алерты выключены по умолчанию
            timeout_seconds=60,
            cache_ttl_seconds=3600,
            params={
                "days": 7,  # период для анализа
                "min_daily_spend": 1.0,  # минимальный средний расход в день ($)
                "min_clicks": 50  # минимум кликов за период
            }
        )

    def get_param_metadata(self) -> Dict[str, Dict[str, Any]]:
        """Возвращает метаданные параметров для UI"""
        return {
            "days": {
                "label": "Период анализа (дней)",
                "description": "Количество дней для анализа кампаний",
                "type": "number",
                "min": 3,
                "max": 365,
                "default": 7
            },
            "min_daily_spend": {
                "label": "Минимальный расход в день ($)",
                "description": "Минимальный средний расход для включения в сегментацию",
                "type": "number",
                "min": 0.5,
                "max": 10000,
                "default": 1.0
            },
            "min_clicks": {
                "label": "Минимум кликов",
                "description": "Минимальное количество кликов за период для включения в анализ",
                "type": "number",
                "min": 10,
                "max": 10000,
                "default": 50
            }
        }

    def analyze(self, config: ModuleConfig) -> Dict[str, Any]:
        """
        Анализ и сегментация кампаний по эффективности.

        Args:
            config: Конфигурация модуля

        Returns:
            Dict[str, Any]: Данные о сегментах кампаний
        """
        # Получение параметров
        days = config.params.get("days", 7)
        min_daily_spend = config.params.get("min_daily_spend", 1.0)
        min_clicks = config.params.get("min_clicks", 50)

        # Период анализа
        date_from = datetime.now().date() - timedelta(days=days - 1)

        # Работа с БД
        with get_db_session() as session:
            # Получаем все кампании с данными за период
            query = session.query(
                Campaign.internal_id,
                Campaign.binom_id,
                Campaign.current_name,
                Campaign.group_name,
                CampaignStatsDaily.date,
                CampaignStatsDaily.clicks,
                CampaignStatsDaily.cost,
                CampaignStatsDaily.revenue,
                CampaignStatsDaily.leads
            ).join(
                CampaignStatsDaily,
                Campaign.internal_id == CampaignStatsDaily.campaign_id
            ).filter(
                CampaignStatsDaily.date >= date_from
            ).order_by(
                Campaign.internal_id,
                CampaignStatsDaily.date
            )

            results = query.all()

            # Группировка по кампаниям
            campaigns_data = defaultdict(lambda: {
                "binom_id": None,
                "name": None,
                "group": None,
                "total_clicks": 0,
                "total_cost": 0,
                "total_revenue": 0,
                "total_leads": 0
            })

            for row in results:
                campaign_id = row.internal_id
                campaigns_data[campaign_id]["binom_id"] = row.binom_id
                campaigns_data[campaign_id]["name"] = row.current_name
                campaigns_data[campaign_id]["group"] = row.group_name or "Без группы"

                campaigns_data[campaign_id]["total_clicks"] += row.clicks or 0
                campaigns_data[campaign_id]["total_cost"] += float(row.cost) if row.cost else 0
                campaigns_data[campaign_id]["total_revenue"] += float(row.revenue) if row.revenue else 0
                campaigns_data[campaign_id]["total_leads"] += row.leads or 0

            # Подготовка кампаний для сегментации
            campaigns = []
            all_rois = []

            for campaign_id, data in campaigns_data.items():
                avg_daily_spend = data["total_cost"] / days

                # Фильтры
                if avg_daily_spend < min_daily_spend:
                    continue
                if data["total_clicks"] < min_clicks:
                    continue
                if data["total_cost"] == 0:
                    continue

                # Расчет метрик
                roi = ((data["total_revenue"] - data["total_cost"]) / data["total_cost"] * 100) if data["total_cost"] > 0 else -100
                cr = (data["total_leads"] / data["total_clicks"] * 100) if data["total_clicks"] > 0 else 0
                profit = data["total_revenue"] - data["total_cost"]

                campaign_info = {
                    "campaign_id": campaign_id,
                    "binom_id": data["binom_id"],
                    "name": data["name"],
                    "group": data["group"],
                    "total_clicks": data["total_clicks"],
                    "total_cost": round(data["total_cost"], 2),
                    "total_revenue": round(data["total_revenue"], 2),
                    "total_leads": data["total_leads"],
                    "profit": round(profit, 2),
                    "avg_daily_spend": round(avg_daily_spend, 2),
                    "roi": round(roi, 1),
                    "cr": round(cr, 2)
                }

                campaigns.append(campaign_info)
                all_rois.append(roi)

            if not campaigns:
                return {
                    "segments": [],
                    "summary": {
                        "total_campaigns": 0,
                        "stars_count": 0,
                        "performers_count": 0,
                        "average_count": 0,
                        "underperformers_count": 0
                    },
                    "quartiles": {},
                    "period": {
                        "days": days,
                        "date_from": date_from.isoformat(),
                        "date_to": datetime.now().date().isoformat()
                    },
                    "params": {
                        "min_daily_spend": min_daily_spend,
                        "min_clicks": min_clicks
                    }
                }

            # Вычисление перцентилей (квартилей)
            sorted_rois = sorted(all_rois)
            q1 = statistics.quantiles(sorted_rois, n=4)[0]  # 25й перцентиль
            q2 = statistics.quantiles(sorted_rois, n=4)[1]  # 50й перцентиль (медиана)
            q3 = statistics.quantiles(sorted_rois, n=4)[2]  # 75й перцентиль

            # Сегментация кампаний
            segments = {
                "stars": [],          # Топ 25% (ROI >= Q3)
                "performers": [],     # 25-50% (Q2 <= ROI < Q3)
                "average": [],        # 50-75% (Q1 <= ROI < Q2)
                "underperformers": [] # Нижние 25% (ROI < Q1)
            }

            for campaign in campaigns:
                roi = campaign["roi"]

                if roi >= q3:
                    segments["stars"].append(campaign)
                elif roi >= q2:
                    segments["performers"].append(campaign)
                elif roi >= q1:
                    segments["average"].append(campaign)
                else:
                    segments["underperformers"].append(campaign)

            # Сортировка сегментов по ROI (убывание)
            for segment_name in segments:
                segments[segment_name].sort(key=lambda x: x["roi"], reverse=True)

            # Создание структурированного вывода
            segments_output = [
                {
                    "segment_name": "Звезды (топ 25%)",
                    "segment_id": "stars",
                    "description": f"ROI > {q3:.1f}%",
                    "campaign_count": len(segments["stars"]),
                    "total_cost": round(sum(c["total_cost"] for c in segments["stars"]), 2),
                    "total_revenue": round(sum(c["total_revenue"] for c in segments["stars"]), 2),
                    "total_profit": round(sum(c["profit"] for c in segments["stars"]), 2),
                    "avg_roi": round(sum(c["roi"] for c in segments["stars"]) / len(segments["stars"]), 1) if segments["stars"] else 0,
                    "campaigns": segments["stars"]
                },
                {
                    "segment_name": "Перформеры (25-50%)",
                    "segment_id": "performers",
                    "description": f"{q2:.1f}% < ROI <= {q3:.1f}%",
                    "campaign_count": len(segments["performers"]),
                    "total_cost": round(sum(c["total_cost"] for c in segments["performers"]), 2),
                    "total_revenue": round(sum(c["total_revenue"] for c in segments["performers"]), 2),
                    "total_profit": round(sum(c["profit"] for c in segments["performers"]), 2),
                    "avg_roi": round(sum(c["roi"] for c in segments["performers"]) / len(segments["performers"]), 1) if segments["performers"] else 0,
                    "campaigns": segments["performers"]
                },
                {
                    "segment_name": "Середнячки (50-75%)",
                    "segment_id": "average",
                    "description": f"{q1:.1f}% < ROI <= {q2:.1f}%",
                    "campaign_count": len(segments["average"]),
                    "total_cost": round(sum(c["total_cost"] for c in segments["average"]), 2),
                    "total_revenue": round(sum(c["total_revenue"] for c in segments["average"]), 2),
                    "total_profit": round(sum(c["profit"] for c in segments["average"]), 2),
                    "avg_roi": round(sum(c["roi"] for c in segments["average"]) / len(segments["average"]), 1) if segments["average"] else 0,
                    "campaigns": segments["average"]
                },
                {
                    "segment_name": "Аутсайдеры (нижние 25%)",
                    "segment_id": "underperformers",
                    "description": f"ROI <= {q1:.1f}%",
                    "campaign_count": len(segments["underperformers"]),
                    "total_cost": round(sum(c["total_cost"] for c in segments["underperformers"]), 2),
                    "total_revenue": round(sum(c["total_revenue"] for c in segments["underperformers"]), 2),
                    "total_profit": round(sum(c["profit"] for c in segments["underperformers"]), 2),
                    "avg_roi": round(sum(c["roi"] for c in segments["underperformers"]) / len(segments["underperformers"]), 1) if segments["underperformers"] else 0,
                    "campaigns": segments["underperformers"]
                }
            ]

            return {
                "segments": segments_output,
                "summary": {
                    "total_campaigns": len(campaigns),
                    "stars_count": len(segments["stars"]),
                    "performers_count": len(segments["performers"]),
                    "average_count": len(segments["average"]),
                    "underperformers_count": len(segments["underperformers"])
                },
                "quartiles": {
                    "q1": round(q1, 1),
                    "q2": round(q2, 1),
                    "q3": round(q3, 1)
                },
                "period": {
                    "days": days,
                    "date_from": date_from.isoformat(),
                    "date_to": datetime.now().date().isoformat()
                },
                "params": {
                    "min_daily_spend": min_daily_spend,
                    "min_clicks": min_clicks
                }
            }

    def generate_recommendations(self, raw_data: Dict[str, Any]) -> List[str]:
        """
        Генерация рекомендаций - ОТКЛЮЧЕНО.
        """
        return []

    def prepare_chart_data(self, raw_data: Dict[str, Any]) -> List[Dict[str, Any]]:
        """Подготовка данных для Chart.js"""
        segments = raw_data.get("segments", [])
        summary = raw_data.get("summary", {})

        if not segments:
            return []

        charts = []

        # График распределения кампаний по сегментам
        charts.append({
            "id": "segmenter_distribution",
            "type": "pie",
            "data": {
                "labels": [s["segment_name"] for s in segments],
                "datasets": [{
                    "data": [s["campaign_count"] for s in segments],
                    "backgroundColor": [
                        "rgba(75, 192, 192, 0.8)",   # Звезды - зеленый
                        "rgba(54, 162, 235, 0.8)",   # Перформеры - синий
                        "rgba(255, 206, 86, 0.8)",   # Середнячки - желтый
                        "rgba(255, 99, 132, 0.8)"    # Аутсайдеры - красный
                    ]
                }]
            },
            "options": {
                "responsive": True,
                "plugins": {
                    "title": {
                        "display": True,
                        "text": "Распределение кампаний по сегментам"
                    },
                    "legend": {
                        "position": "bottom"
                    }
                }
            }
        })

        # График прибыли по сегментам
        charts.append({
            "id": "segmenter_profit",
            "type": "bar",
            "data": {
                "labels": [s["segment_name"] for s in segments],
                "datasets": [{
                    "label": "Прибыль ($)",
                    "data": [s["total_profit"] for s in segments],
                    "backgroundColor": [
                        "rgba(75, 192, 192, 0.6)" if p >= 0 else "rgba(255, 99, 132, 0.6)"
                        for p in [s["total_profit"] for s in segments]
                    ],
                    "borderColor": [
                        "rgba(75, 192, 192, 1)" if p >= 0 else "rgba(255, 99, 132, 1)"
                        for p in [s["total_profit"] for s in segments]
                    ],
                    "borderWidth": 1
                }]
            },
            "options": {
                "responsive": True,
                "plugins": {
                    "title": {
                        "display": True,
                        "text": "Прибыль по сегментам"
                    }
                },
                "scales": {
                    "y": {
                        "beginAtZero": True,
                        "title": {
                            "display": True,
                            "text": "Прибыль ($)"
                        }
                    }
                }
            }
        })

        return charts

    def generate_alerts(self, raw_data: Dict[str, Any]) -> List[Dict[str, Any]]:
        """
        Генерация алертов для сегментации кампаний.
        """
        segments = raw_data.get("segments", [])
        alerts = []

        # Находим аутсайдеров (нижний сегмент)
        underperformers = next((s for s in segments if s["segment_id"] == "underperformers"), None)
        if underperformers and underperformers["campaign_count"] > 0:
            total_loss = abs(underperformers["total_profit"]) if underperformers["total_profit"] < 0 else 0

            if total_loss > 0:
                message = f"Обнаружено {underperformers['campaign_count']} кампаний в сегменте Аутсайдеров\n\n"
                message += f"Общий убыток сегмента: ${total_loss:.2f}\n"
                message += f"Средний ROI: {underperformers['avg_roi']:.1f}%\n\n"
                message += "Топ-3 худших кампании:\n"

                for i, campaign in enumerate(underperformers["campaigns"][:3], 1):
                    message += f"{i}. {campaign['name']}: ROI {campaign['roi']:.1f}%, убыток ${abs(campaign['profit']):.2f}\n"

                alerts.append({
                    "type": "segmenter_underperformers",
                    "severity": "high",
                    "message": message,
                    "recommended_action": "Рассмотрите возможность остановки или оптимизации этих кампаний",
                    "campaigns_count": underperformers["campaign_count"],
                    "total_loss": round(total_loss, 2)
                })

        return alerts
